Tutorial documentation (#2)
AgPipeline/Atlana: docs/index.md

Atlana is our web interface used to configure and run workflows for processing drone captured data.

## What's available

> The UI provides a convienient way for running workflows.

Reviewer: ```suggestion

The UI provides a convienient way for running workflows.
```

It also provide ways to:

 * configure one or more storage locations at the folder and file levels
 * download and restore configured storage locations
 * run workflows
 * create new workflows that use git repositories
 * re-run jobs you've created and jobs others have created
 * download and restore workflows you've created and workflows you've run

The downloaded configurations of storage locations and workflows are in JSON format and can be shared with others.

## Tutorial

> Items needed:

Reviewer: ```suggestion

Required inputs (or see [sample data](#sample-data):
```

- an orthomosaic image to be processed
- a GeoJSON file containing the plots of interest
- an optional YAML file containing [experiment information](https://osf.io/xdkcy/wiki/Configuration%20YAML/) [![external link](img/external_link.png)](https://osf.io/xdkcy/wiki/Configuration%20YAML/)

> What we will  be doing:

Reviewer: ```suggestion
What we will do:
```

> 1. [Upload files and configure](#uploading-and-configuring-files) a server-side file storage definition

Reviewer: unclear jargon - what is 'server-side' and what is a 'file storage definition'

> 2. Running the [Canopy Cover image-based workflow](#running-a-workflow)

Reviewer: ```suggestion
2. Run the [Canopy Cover workflow](#running-a-workflow)
```

> 3. [Downloading the results](#downloading-the-results) to the local computer

Reviewer: ```suggestion
3. [Download results](#downloading-the-results) to the local computer
```

> 4. [Deleting](#delete-workflow-run) the workflow run

Reviewer: ```suggestion
4. [Delete](#delete-workflow-run) the workflow run
```

(why??)

Note that the geographic boundaries of the image don't need to match that of the plots.
Only those plots that intersect the image will be processed.

Additionally, the coordinate system of the image doesn't need to match that of the plots.
Plot points are converted automatically to the image's geographic coordinate system.

### Sample data

We also provide an [archive of sample data](https://data.cyverse.org/dav-anon/iplant/projects/aes/cct/diag/sample-data/sample_website_data.tar.gz
) [![external link](img/external_link.png)](https://data.cyverse.org/dav-anon/iplant/projects/aes/cct/diag/sample-data/sample_website_data.tar.gz) that can be downloaded and used in this Tutorial.
Be sure to extract the files from this archive before uploading them.

> Uploaded archives are **not** usable by the system.

Reviewer: ```suggestion
```

removed for brevity (hopefully the system will let the user know)

Author: I would like to keep the sentence. Users find out when the file they want doesn't exist on the server. I feel it's better to warn them ahead of time. Issue: https://github.com/AgPipeline/issues-and-projects/issues/559

The following `bash` commands can be used to download the archive and extract the files.

```bash
curl --output sample_website_data.tar.gz https://data.cyverse.org/dav-anon/iplant/projects/aes/cct/diag/sample-data/sample_website_data.tar.gz
tar -xzf sample_website_data.tar.gz
```

### Uploading and configuring files

On the main page of the site, click the `files` link under **Data Sources**.
This will take you to the file storage definition page

<img src="img/landing_click_files.png" width="450" alt="File storage definition link" aria-label="File storage definition link" title="Click on Files link"/>

> Next click the upload icon, select and upload the files to process.

Reviewer: the upload icon is too faint; I didn't notice it at first

Author: Issue: https://github.com/AgPipeline/issues-and-projects/issues/560

You may see an <img src="img/configure_files_uploading_files.png" width="90" alt="Uploading files" aria-label="Uploading files" title="Uploading files"/> message on the icon if it takes a few seconds to upload the files.
You can also drag-and-drop files onto the upload location.

<img src="img/configure_files_upload.png" width="450" alt="Click upload files icon" aria-label="Upload files link" title="Click or drag-and-drop to upload files"/>

> Next, select "Server Side" from the drop down on the upper right of the page, and click "New".

Reviewer: it is unclear what is happening here - 
- after I upload, I don't see a list of the files 
- not clear what 'server side' / new should do - is this just setting the destination? Would it be possible to set the destination at the time of upload?

Author: > after I upload, I don't see a list of the files

I added a line about seeing the uploaded files on the next step

> not clear what 'server side' / new should do - is this just setting the destination? Would it be possible to set the destination at the time of upload?

At the start of the section I've added some more text explaining what we're doing. I've also added clarification on the line

Reviewer: I think that "Server side" is too jargony. There must be a more common term for what this is

Author: Issue written to change this: https://github.com/AgPipeline/issues-and-projects/issues/564

<img src="img/configure_files_server_side.png" width="450" alt="Choose dropdown server-side then new" aria-label="Create new server-side definition" title="Create new server-side file configuration"/>

This will display a new window showing the uploaded files.

> Change the name of the configuration if you want to.

Reviewer: not clear - why would I want to change the name of the configuration? what _is_ the configuration?

(update - it seems like 'configuration' is the name of a set of input files - would it make sense to call this a 'project'?

Author: > (update - it seems like 'configuration' is the name of a set of input files - would it make sense to call this a 'project'?

I think the term "project" would be confusing since it's just a place to put files. I'm using the term `endpoint` now (once the updates are uploaded)

Reviewer: 'endpoint' seems too esoteric. How about 'input folder' or 'project folder'?

Author: Reworded the 'endpoint' term to be related to the action of defining a folder on the server

> Select "OK" to complete the configuration and dismiss the window.

Reviewer: when I try to rename the configuration I get an error "duplicate name found"

Author: @dlebauer I am unable to reproduce this. Can you provide the steps needed?

Reviewer: - Create a folder holding a set of files
- Create a new one. Give it the same name as the one that already exists
- Error shows up

But this isn't really a problem. Will mark as resolved

<img src="img/configure_files_new_server_side.png" width="450" alt="Finish server-side configuration" aria-label="Click OK to complete configuration" title="Click OK to accept"/>

The configured access is now available to workflows and shown on the page.

<img src="img/configure_files_completed.png" width="450" alt="Configured file storage listing" aria-label="Configured file storage listing" title="Configured file storage listing"/>

You can return to the main page by clicking the "back" button.

<img src="img/configure_files_back.png" width="450" alt="Click back button" aria-label="Click back button for main page" title="Click the back button"/>

### Running a workflow

> On the main page of the site, click the `Image` link under **Workflows**.

Reviewer: calling this 'Image' is confusing, especially since above 'image' refers to a picture as in 'orthomosaic image'. Perhaps call it an algorithm, or a workflow step?

Author: I'm trying to keep Image-based workflows separate from Lidar-based ones, etc.

Reviewer: ok got it - I thought that this was 'image' as in 'docker image'

I would refer to these as 'RGB Image' as 'Image' could refer to multi/hyperspectral image, xray image, etc

Author: I am adding an issue for this clarification: https://github.com/AgPipeline/issues-and-projects/issues/555

Choose "Canopy Cover" from the drop down on the upper right of the page, and click "Run".

<img src="img/workflows_image_cc_run.png" width="450" alt="Choose dropdown Canopy Cover then run" aria-label="Run Canopy Cover workflow" title="Run Canopy Cover workflow"/>

The fields needed for running this workflow are shown on the page.
We will be configuring the "Image file" field next.

<img src="img/workflows_image_cc_fields.png" width="450" alt="Canopy Cover workflow fields" aria-label="Available Canopy Cover workflow fields" title="Canopy Cover workflow fields"/>

Select the ellipses next to "Image file" to bring up the browsing window.
Since we've  only defined one file configuration, it will be automatically chosen for us.
Select the orthomosaic file and click "OK" to confirm your selection and dismiss the window.
If you are using the sample data, the name of the file to choose is "sample_orthophoto.tif".

<img src="img/workflows_image_cc_field_file.png" width="450" alt="Selecting image to process" aria-label="File selection window" title="Select image to process"/>

> Click the ellipses next to each of the other entries to choose the GeoJSON and YAML experiment files for this workflow.

Reviewer: can this also take a shapefile? does that require multiple shape files be zipped?

Author: We made a decision a while ago to only accept GeoJSON files. I'm not sure, are you suggesting that we provide shapefile->GeoJSON functionality on the UI?

Reviewer: >  decision a while ago to only accept GeoJSON files

Sorry if this didn't come across correctly - at the level of an algorithm, yes, it should only accept 'standard' data formats like geojson. _but_ having another tool that converts shp --> geojson to 'help users out' would also be helpful. At the UI level it might make sense to have this happen behind the scenes automatically, although this is also a chance to let them know 'that is not acceptable - let us convert it to an acceptable format for you'

Author: I will add an issue to add this functionality: https://github.com/AgPipeline/issues-and-projects/issues/556

This will display a page where the status of the job is display and you can access normal and warning/error messages.
To update the messages and warnings/errors, click the "Refresh" button.
If there are many messages retrieved from the server, it may take a few seconds before the messages are displayed.

The current count of regular and warning/error messages are updated after refreshing.
Clicking on the counts will switch between the message types.

<img src="img/workflows_image_cc_messages.png" width="450" alt="Workflow output listing refresh" aria-label="Viewing output messages" title="Refresh view of output messages"/>

Use the "back" button to return to the workflow listing page.
Click the "View" button again on the workflow listing page to re-display the messages.

<img src="img/workflows_image_cc_messages_back.png" width="450" alt="Back button to view workflows" aria-label="Click back button for workflow listing" title="Click the back button"/>

### Downloading the results

On the workflow listing page, completed workflows have a download button to the far right of their name.

> Clicking the download button will download the CSV file ccontaining the results and the JSON of the job definition.

Reviewer: canopy_cover.csv looks great!

canopy_cover.json contains:

```
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<title>500 Internal Server Error</title>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.</p>
```

is there a way to download the clipped and masked images?

Author: Canopy cover JSON file should not be doing that; I'll look into it. There isn't a way to get everything downloaded - I'll put in an issue

Reviewer: >  get everything downloaded 

It should still be an option to download the intermediate files - some people may just want the CSV files.

Author: Issues written up:
https://github.com/AgPipeline/issues-and-projects/issues/552
https://github.com/AgPipeline/issues-and-projects/issues/557
